crema/parsers/msfragger.py

```diff
@@ -0,0 +1,106 @@
+"""A parser for the MSFragger tab-delimited format"""
+import re
+import logging
+
+import pandas as pd
+
+from .txt import read_txt
+from .pepxml import read_pepxml
+from .. import utils
+
+LOGGER = logging.getLogger(__name__)
+
+
+def read_msfragger(txt_files, pairing_file_name=None, copy_data=True):
+    """Read peptide-spectrum matches (PSMs) from MSFragger tab-delimited files.
+
+    Parameters
+    ----------
+    txt_files : str, pandas.DataFrame or tuple of str
+        One or more collection of PSMs in the MSFragger tab-delimited format.
+    pairing_file_name : str, optional
+        A tab-delimited file that explicity pairs target and decoy peptide
+        sequences. Requires one column labled 'target' that contains target
+        sequences and a second colun labeled 'decoy' that contains decoy
+        sequences.
+    copy_data : bool, optional
+        If true, a deep copy of the data is created. This uses more memory, but
+        is safer because it prevents accidental modification of the underlying
+        data. This argument only has an effect when `txt_files` is a
+        :py:class:`pandas.DataFrame`
+
+    Returns
+    -------
+    PsmDataset
+        A :py:class:`~crema.dataset.PsmDataset` object containing the parsed
+        PSMs.
+    """
+    target = "label"
+    peptide = "peptide"
+    spectrum = ["Filename", "start scan"]
+    pairing = ""
+    protein = "protein"
+    protein_delim = ";"
+    decoy_prefix = "rev_"
+
+    # Possible score columns output by MSFragger.
+    scores = {
+        "hyperscore",
+        "nextscore",
+        "expect",
+        "expectscore",
+    }
+    # scores_all = scores
+
+    # TODO need to figure out how to for this
+    # Keep only crux scores that exist in all of the files.
+    # if isinstance(txt_files, pd.DataFrame):
+    #    scores = scores.intersection(set(txt_files.columns))
+    # else:
+    #    txt_files = utils.listify(txt_files)
+    #    for txt_file in txt_files:
+    #        with open(txt_file) as txt_ref:
+    #            cols = txt_ref.readline().rstrip().split("\t")
+    #            scores = scores.intersection(set(cols))
+
+    # if not scores:
+    #    raise ValueError(
+    #        "Could not find any of the MSFragger score columns in all of the files."
+    #        f"The columns crema looks for are {', '.join(list(scores_all))}"
+    #    )
```

Comment:
**comment:** There isn't a way to do this without parsing through the whole file first. We would need to keep track of the scores from each parsed file and check the intersection either at the end or after each to see that it remains `> 0`.

Reply:
Ok - I have edited this script to account for this.